skbio/tree/_tree.py

```diff
@@ -2705,6 +2705,47 @@ def compare_tip_distances(self, other, sample=None, dist_f=distance_from_r,
         return dist_f(self_matrix, other_matrix)
 
     @experimental(as_of="0.4.0")
+    def bifurcate(self, insert_length=None, remove_singles=True):
+        r""" Restructures tree into a bifurcating tree
+
+        All nodes that have more than 2 children will
+        have additional intermediate nodes injected to ensure that
+        every node has only 2 children
+        Also, any nodes that have a single child will be collapsed
+
+        Parameters
+        ----------
+        remove_singles : bool, optional
+            Removes all nodes that have only a single child and relink
+            of their childrent to their parent node
+
+        Examples
+        --------
+        >>> from skbio import TreeNode
+        >>> from six import StringIO
+        >>> tree = TreeNode.read(StringIO("((a,b,g,h)c,(d,e)f)root;"))
+        >>> tree.bifurcate()
+        >>> print(tree)
+        ((h,(g,(a,b)))c,(d,e)f)root;
+        <BLANKLINE>
+
+        """
+        for n in self.traverse(include_self=True):
+            if len(n.children) > 2:
+                stack = n.children
+                while len(stack) > 2:
+                    ind = stack.pop()
```

Comment:
I'm basically creating a stack for all of the children and reordering the children if there are more than 2 children.  

The references shouldn't change since this is an auxiliary data structure that isn't explicitly touching the tree itself.  As far as I can tell, it isn't a problem considering the tests are behaving as expected.  Do you have suggestions on how to test this?